realm/realm-library/src/objectServer/java/io/realm/internal/objectstore/OsRemoteMongoCollection.java

```diff
@@ -38,12 +65,231 @@ public long getNativeFinalizerPtr() {
         return nativeFinalizerPtr;
     }
 
-    public void count(String filter) {
-        throw new UnsupportedOperationException("Not Implemented");
+    public Long count() {
+        return count(null);
+    }
+
+    public Long count(@Nullable final Bson filter) {
+        return count(filter, null);
+    }
+
+    public Long count(@Nullable final Bson filter, @Nullable final RemoteCountOptions options) {
+        AtomicReference<Long> success = new AtomicReference<>(null);
+        AtomicReference<ObjectServerError> error = new AtomicReference<>(null);
+        OsJNIResultCallback<Long> callback = new OsJNIResultCallback<Long>(success, error) {
+            @Override
+            protected Long mapSuccess(Object result) {
+                return (Long) result;
+            }
+        };
+
+        // no filter means count all
+        String filterString = filter == null ?
+                JniBsonProtocol.encode(new Document(), codecRegistry) :
+                JniBsonProtocol.encode(filter, codecRegistry);
+        int limit = options == null ? 0 : options.getLimit();
+
+        nativeCount(nativePtr, filterString, limit, callback);
+
+        return ResultHandler.handleResult(success, error);
+    }
+
+    public DocumentT findOne() {
+        return findOne(new Document());
+    }
+
+    public <ResultT> ResultT findOne(final Class<ResultT> resultClass) {
+        AtomicReference<ResultT> success = new AtomicReference<>(null);
+        AtomicReference<ObjectServerError> error = new AtomicReference<>(null);
+        OsJNIResultCallback<ResultT> callback = new OsJNIResultCallback<ResultT>(success, error) {
+            @Override
+            protected ResultT mapSuccess(Object result) {
+                return (ResultT) findOneSuccessMapper(result, resultClass);
+            }
+        };
+
+        nativeFindOne(nativePtr, JniBsonProtocol.encode(new Document(), codecRegistry), callback);
+
+        return ResultHandler.handleResult(success, error);
+    }
+
+    public DocumentT findOne(final Bson filter) {
+        AtomicReference<DocumentT> success = new AtomicReference<>(null);
+        AtomicReference<ObjectServerError> error = new AtomicReference<>(null);
+        OsJNIResultCallback<DocumentT> callback = new OsJNIResultCallback<DocumentT>(success, error) {
+            @Override
+            protected DocumentT mapSuccess(Object result) {
+                return (DocumentT) findOneSuccessMapper(result, Document.class);
```

Comment:
This looks wrong? `DocumentT` can be any type, so what ensures this cast is correct?

Reply:
That's right. The return type in `findOneSuccessMapper` was wrongly specified. I've fixed it there so the typecasting isn't necessary anymore.